components/proto/src/funder/messages.rs

```diff
@@ -313,7 +313,16 @@ impl FriendsRoute {
     /// A -- B -- C -- D -- E -- F -- A   (Single cycle, first == last)
     /// A -- B -- C -- D -- E -- F        (A route with no repetitions)
     pub fn is_valid(&self) -> bool {
-        if (self.public_keys.len() < 2) || (self.public_keys.len() > MAX_ROUTE_LEN) {
+        if self.public_keys.len() < 2 {
```

Comment:
I agree with you on the `is_valid()` and `is_valid_part()` split.
From what I noticed: 
- `is_valid()` is only used when the user composes a new transaction (in handle_control.rs).
- `is_valid_part()` is used when receiving or sending a request (handle_friend.rs, incoming.rs, outgoing.rs).

### Expected invariants

```rust
is_valid(A,B,C,D) == true // usual route
is_valid(A,B,C,D,A) == true // cyclic route that is at least 3 nodes long, having first item equal the last item
is_valid(A,A) == false // cyclic route that is too short (only 2 nodes long)
is_valid(A,B,C,B,D) == false // Should have no repetitions that are not the first and last nodes.

is_valid_part(A,B,C,D) == true
is_valid_part(A,B,C,D,A) == false // should have no cycles in a partial route
is_valid_part(A,A) == false // should have no repetitions ins a partial route
is_valid_part(A,B,C,B,D) == false // should have no repetitions in a partial route
```

To be sure about your implementation, I think you can add those as tests.

Currently `is_valid_part(A,B,C,D,A) == true`, which I think is incorrect.
Also note that `is_valid(A,A) == true` at the old implementation, which I also believe to be incorrect. I only noticed it now.


### Why are routes complicated?

There is some use for cyclic routes: They can allow a node to rebalance his debts, possibly moving his debts from one entity to another. This is done by letting a node pay himself along a cyclic route.

Generally I would have prefer to have no conditions at all. (I also believe that the mentioned routes rules are already enforced by the rest of the system). However, I have this fear that I might have missed something and this would cause a security problem. It's difficult to me to think about a real use for a strange route that does multiple cycles. (Maybe there is though).

In addition, if something is not possible I would like users to know this as soon as possible. Therefore we already check the validity of a route when the transaction is created. This is a "client side" check for a route that will probably be rejected anyway in the future.